### PR TITLE
[compute/cker] Introduce the ShapeIterator

### PR DESCRIPTION
**Pull request by tomdol**
This commit introduces a utility that effectively makes the Shape objects iterable. It's an iterator class which points to the individual dimensions in the shape and allows the interoperability of the Shape class and STL algorithms as well as range-based for loops. The iterator fulfills the requirements of a bidirectional iterator.

In addition this commit contains one extra utility which allows the Shape objects conversion to std::string by contatenating them with a comma.

ONE-DCO-1.0-Signed-off-by: Tomasz Dolbniak <t.dolbniak@partner.samsung.com>

### CODE REVIEW

**Comment by zetwhite** [comment]
@nnfw-bot onert-x64-release

**Comment by zetwhite** [comment]
@nnfw-bot test onert-x64-release

**Comment by tomdol** [comment]
@zetwhite could you please review or suggest particular reviewers for this PR?

**Comment by hseok-oh** [comment]
@tomdol 
1) What is purpose of these utilities? Where will be it used?
2) Why `neg_ShapeIterator_empty_shape` have negative test naming?

**Comment by tomdol** [comment]
@hseok-oh the purpose is to extend the ability to iterate over the Shape dimensions. With the Shape class' API it's only possible to use it with traditional indexed loops and the `Shape::Dims(int)` method to access individual dimensions. The shape iterator adds the ability to iterate over a shape with a range based for loop:

```
Shape shape{1,2,3,4};
for (auto&& dimension : shape) {}
```

or with algorithms expecting a pair of iterators denoting a range of dimensions:

```
Shape shape{1,2,3,4};
const auto first = begin(shape);
const auto last = end(shape);
std::vector<int> filtered_dims;
// copy the dimensions that are different than 1
std::copy_if(first, last, begin(filtered_dims), [](int dimension) { return dimension != 1; });
// or:
// accumulate all dimensions to get the numer of elements in a tensor
// or:
// get the iterator to the first dimension, skip over the batch and channels and process the remaining dimensions using STL
```

It's difficult to list all possible use cases but the main thing I was missing was the ability to use STL algorithms when traversing a Shape object.

Regarding the test naming, I know there's a requirement for the number of negative tests but couldn't really come up with a good negative test for this utility. Should that be changed in any way?

File: compute/cker/include/cker/ShapeIterator.h [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __NNFW_CKER_SHAPE_ITERATOR_H__
+#define __NNFW_CKER_SHAPE_ITERATOR_H__
+
+#include <utility>
+#include <iostream>
```
Comment (zetwhite) [inline compute/cker/include/cker/ShapeIterator.h:21]: This header looks unnecessary. 
```suggestion
```

**Review by hseok-oh — APPROVED** [review]
LGTM

**Review by zetwhite — APPROVED** [review]

**Pull request merged**